datafusion/core/src/datasource/physical_plan/finder.rs

```diff
@@ -0,0 +1,114 @@
+//! partitioned file sources.
+
+use std::sync::Arc;
+
+use datafusion_common::{
+    tree_node::{TreeNode, VisitRecursion},
+    Result,
+};
+
+use crate::{datasource::listing::PartitionedFile, physical_plan::ExecutionPlan};
+
+use super::{AvroExec, CsvExec, NdJsonExec, ParquetExec};
+
+pub type FinderFunction =
+    Box<dyn Fn(&dyn ExecutionPlan) -> Option<Vec<Vec<PartitionedFile>>>>;
+
+/// Get all of the [`PartitionedFile`] to be scanned for an [`ExecutionPlan`]
+///
+/// This structure will find all `ScanFileConfig` in any built in
+/// [`ExecutionPlan`] and allows finding user defined nodes as well
+pub struct PartitionedFileFinder {
+    custom_finder: FinderFunction,
+}
+
+impl Default for PartitionedFileFinder {
+    fn default() -> Self {
+        Self {
+            // default custom finding function does nothing
+            custom_finder: Box::new(|_plan| None),
+        }
+    }
+}
+
+impl PartitionedFileFinder {
+    /// Create a new file finder
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Get all [`PartitionedFile`]s that are scanned for an
+    /// [`ExecutionPlan`],  by recursively checking all children
+    pub fn find(&self, plan: Arc<dyn ExecutionPlan>) -> Vec<Vec<Vec<PartitionedFile>>> {
+        let mut collector: Vec<Vec<Vec<PartitionedFile>>> = vec![];
+        plan.apply(&mut |plan| {
+            if let Some(files) = self.get_files(plan.as_ref()) {
+                collector.push(files);
+            }
+            Ok(VisitRecursion::Continue)
+        })
+        .expect("infallable");
+        collector
+    }
+
+    /// Provide a custom method to find `PartitionedFiles` for
+    /// `ExecutionPlans`
+    ///
+    /// Called on all [`ExecutionPlan`]s other than built ins such as
+    /// [`ParquetExec`] and can be used to extract
+    /// [`PartitionedFile`]s from user defined nodes
+    pub fn with_finder<F>(mut self, custom_finder: F) -> Self
+    where
+        F: Fn(&dyn ExecutionPlan) -> Option<Vec<Vec<PartitionedFile>>> + 'static,
+    {
+        self.custom_finder = Box::new(custom_finder);
+        self
+    }
+
+    /// Return the [`PartitionedFile`] scanned by this plan node, or
+    /// `None` if the plan does not can files
+    fn get_files(&self, plan: &dyn ExecutionPlan) -> Option<Vec<Vec<PartitionedFile>>> {
+        let plan_any = plan.as_any();
+        if let Some(parquet_exec) = plan_any.downcast_ref::<ParquetExec>() {
+            Some(parquet_exec.base_config().file_groups.clone())
+        } else if let Some(avro_exec) = plan_any.downcast_ref::<AvroExec>() {
+            Some(avro_exec.base_config().file_groups.clone())
+        } else if let Some(json_exec) = plan_any.downcast_ref::<NdJsonExec>() {
+            Some(json_exec.base_config().file_groups.clone())
+        } else if let Some(csv_exec) = plan_any.downcast_ref::<CsvExec>() {
+            Some(csv_exec.base_config().file_groups.clone())
+        } else {
+            (self.custom_finder)(plan)
+        }
```

Comment:
I agree it is ugly 

However, the API in this PR is extensible in the sense that it can handle any new user defined execution plan (the test contains an example of doing so). 

Also I don't think this PR prevents feature gating, as @tustvold  explains in https://github.com/apache/arrow-datafusion/issues/7357#issuecomment-1707193574

However it is ugly, as you say

Reply:
There are two other solutions:

use a trait->trait casting helper crate, see https://docs.rs/cast_trait_object/latest/cast_trait_object/ and the "Alternatives" section in that doc.

Or make `fn file_groups(&self) -> ...` a member of `ExecutionPlan` and default it to `None` within the trait, so not everybody has to define it.

Reply:
I think we should try avoid adding dependencies if we can help it and your second suggestion doesn't solve anything since the return type would still be a dependency (note that the `file_scan_config` already has a default impl returning `None` ... this is not the issue ... the issue is the type dependency).

But I think I've already found a nice solution in #7485.

Reply:
> But I think I've already found a nice solution in https://github.com/apache/arrow-datafusion/pull/7485.

Mmh, that just flips the issue. Now every exec node has to know every relevant metadata collector, meaning the latter isn't extendable anymore. Actually I think the metadata collector shouldn't be a relevant interface at all. It's the metadata / extensible property that counts.

How about the following interface:

```rust
trait ExecutionPlan {
   ...

   fn get_metadata(&self, type_id: TypeId) -> Option<Box<dyn Any>>;
}

/// Helper
///
/// This is required to keep [`ExecutionPlan`] object-safe.
pub fn get_metadata<T>(plan: &dyn ExecutionPlan) -> Option<T> {
    plan.get_metadata(TypeId::of::<T>()).map(|x| *x.downcast::<T>().expect("wrong type returned"))
}
```

Reply:
> Now every exec node has to know every relevant metadata collector, meaning the latter isn't extendable anymore.

Your example code is not any more extensible than what I wrote ... the method implementation still has to know every relevant type. The only difference is that your code could panic when a method implementation returns a wrong type.

Reply:
Maybe we can meet in the middle: instead of defining freestanding extractor types, we could define `Extractor<T>` where `T` is the metadata that the API user actually wants?

Reply:
Taking a step back, the get_scan_files function appears to only be used in a single test, perhaps we could simply remove it?

Reply:
@crepererum I don't really see how an `Extractor<T>` type would improve anything. It would just add a difficult to understand type for a very niche use case to the public API. And I don't think it would improve the ergonomics at all ... since the extraction logic may very well depend on the type being extracted (and the API I suggest nicely let's you [reuse such logic](https://github.com/apache/arrow-datafusion/blob/4a1c7abb1ace13a42a03375d4949a1260d4cd10b/datafusion/core/src/datasource/physical_plan/file_scan_config.rs#L81-L83)).

@tustvold Right I also noticed that it's not really used. It was apparently introduced in #5572 to address #5566:

> Currently the TreeNodeRewriter is as a visitor to transform a node to another. However, sometimes we don't need to do the transformation and what we want is only to collect some info from the node. To achieve this, it's better to introduce another visitor for collecting info and keep the node unchanged.

I'm not familiar enough with datafusion to determine whether or not that motivation still applies.

Reply:
Proposal in https://github.com/apache/arrow-datafusion/pull/7487